### ISSUE TITLE
Make listchars extends/precedes, fillchars lastline, and "<<<" consistent

### ISSUE THREAD

**Issue by Asheq** [vim/vim#15134]
# Current Behavior

When `wrap` is off, the characters used to show that there is text before or after the visible part of the buffer line come from the `listchars` flags "precedes" and "extends", respectively:

```
:set nowrap
:set listchars=precedes:◁,extends:▷
:set list
```

<img width="640" alt="image" src="https://github.com/vim/vim/assets/4657140/f3186d02-534e-4daa-8e74-eb84851389f8">

---

When `wrap` is on, the character used to show that there is text AFTER the visible part of the buffer line is "@@@" but can be overridden by the`fillchars` flag "lastline". And if `smoothscroll` is on, the character used to show that there is text BEFORE the visible part of the buffer line is "<<<" and cannot be overridden.

```
:set nolist
:set wrap
:set display=lastline
:set smoothscroll
```

<img width="640" alt="image" src="https://github.com/vim/vim/assets/4657140/46b024d2-20e5-4a8a-97d3-4d7453387dd0">

```
:set fillchars=lastline:➤
```

<img width="640" alt="image" src="https://github.com/vim/vim/assets/4657140/080a7c20-b0c8-456a-810b-6c16b0d6611a">

---

And if `showbreak` is not empty, then "<<<" is not displayed.

```
:set showbreak=➥
```

<img width="640" alt="image" src="https://github.com/vim/vim/assets/4657140/7c0e24c3-5c10-4ad5-8cc0-67e047acba92">

---

And if `list` is on and `listchars` includes the "precedes" flag, then the "precedes" character is drawn on top of the `showbreak` character.

```
:set list
```

<img width="640" alt="image" src="https://github.com/vim/vim/assets/4657140/1b01fcd5-a234-4870-ace9-9f6d0f5b5656">

---

# Desired Behavior

- `listchars` flag "precedes" should not be used when `wrap` is on, the same way that the `listchars` flag "extends" is not used when `wrap` is on.
- The default "@@@" should be ">>>" to match the default "<<<" on the other side. It should also be override-able by a new `fillchars` flag called "firstline" to be the counterpart to the existing `lastline` flag.

**Comment by zeertzjq** [comment]
> - 'listchars' flag `precedes` should not be used when 'wrap' is on. The same way that `extends` is not used when 'wrap' is on.

This is not desirable and not backward-compatible. However if a 'fillchars' `firstline` flag is added, it may make sense to give it higher priority over 'listchars' `precedes` flag.

**Comment by Asheq** [comment]
> > * 'listchars' flag "precedes" should not be used when 'wrap' is on.
> 
> This is not desirable and not backward-compatible. However if a 'fillchars' "firstline" flag is added, it may make sense to give it higher priority over 'listchars' "precedes" flag.

Why is `precedes` used on the left side of the buffer line, but `extends` is not used on right side? It appears to be an oversight.

To make things consistent, we can either:
- Stop using `precedes` on the left side OR
- Start using `extends` on the right side

**Comment by zeertzjq** [comment]
> > > * 'listchars' flag "precedes" should not be used when 'wrap' is on.
> > 
> > 
> > This is not desirable and not backward-compatible.
> 
> Why is `precedes` used on the left side of the buffer line, but `extends` is not used on right side? It appears to be an oversight, not something we should intentionally preserve.

It is intentional. Please see [patch 8.1.2060](https://github.com/vim/vim/commit/bffba7f7042f6082e75b42484b15f66087b01941).

**Comment by Asheq** [comment]
> it may make sense to give it higher priority over 'listchars' "precedes" flag.

The priority chain would get even more odd I think:
1. default "<<<" (lowest priority)
2. showbreak
3. 'listchars' "precedes" flag
4. 'fillchars' "firstline" flag (highest priority)

If we continue using 'listchars' "precedes",  I propose we change the order:
1. default "<<<" (lowest priority)
2. 'listchars' "precedes" flag
3. 'fillchars' "firstline" flag
4. showbreak (highest priority)

If we stop using 'listchars' "precedes", it would be:
1. default "<<<" (lowest priority)
2. 'fillchars' "firstline" flag
3. showbreak (highest priority)

**Comment by Asheq** [comment]
> > > > * 'listchars' flag "precedes" should not be used when 'wrap' is on.
> > > 
> > > 
> > > This is not desirable and not backward-compatible.
> > 
> > 
> > Why is `precedes` used on the left side of the buffer line, but `extends` is not used on right side? It appears to be an oversight, not something we should intentionally preserve.
> 
> It is intentional. Please see [patch 8.1.2060](https://github.com/vim/vim/commit/bffba7f7042f6082e75b42484b15f66087b01941).

I also see [patch 9.0.0758](https://github.com/vim/vim/commit/13cdde39520220bb856cba16626327c706752b51)

I just feel that:
- Either those changes should not have been made OR
- They should have been accompanied by similar changes for "extends"

**Comment by zeertzjq** [comment]
The situation with 'showbreak' is a bit different. 'listchars' "precedes" is drawn over 'showbreak', and it doesn't disable 'showbreak' (which can be seen if 'showbreak' contains multiple characters), while 'showbreak' disables the default "<<<". If 'fillchars' "firstline" flag is added, it makes sense for it to draw over 'showbreak' as well.

**Comment by Asheq** [comment]
> The situation with 'showbreak' is a bit different. 'listchars' "precedes" is drawn over 'showbreak', and it doesn't disable 'showbreak' (which can be seen if 'showbreak' contains multiple characters), while 'showbreak' disables the default "<<<". If 'fillchars' "firstline" flag is added, it makes sense for it to draw over 'showbreak' as well.

I don't think it makes sense for "precedes" to draw over `showbreak` characters. It doesn't give the user any more information than what `showbreak` already does. It just makes the UI more cluttered. That's why I proposed `showbreak` being the highest priority. If showbreak is non-empty, nothing else needs to be shown.

As far as the proposed `fillchars` "firstline" goes, I think it should actually have the same priority as the default "<<<". It's simply changing the default character, not adding another entry in the priority list. IOW, if `showbreak` disables the default "<<<", `fillchars` "firstline" should not show either. I hope that makes sense.
